clients/src/main/java/org/apache/kafka/clients/producer/internals/ProduceRequestResult.java

```diff
@@ -41,12 +43,20 @@ public ProduceRequestResult() {
      * Mark this request as complete and unblock any threads waiting on its completion.
      * @param topicPartition The topic and partition to which this record set was sent was sent
      * @param baseOffset The base offset assigned to the record
+     * @param responseTimestamp The timestamp returned by the broker
      * @param error The error that occurred if there was one, or null.
      */
-    public void done(TopicPartition topicPartition, long baseOffset, RuntimeException error) {
+    public void set(TopicPartition topicPartition, long baseOffset, long responseTimestamp, RuntimeException error) {
         this.topicPartition = topicPartition;
         this.baseOffset = baseOffset;
+        this.responseTimestamp = responseTimestamp;
         this.error = error;
+    }
+
+    /**
+     * Mark this request as complete and unblock any threads waiting on its completion.
+     */
+    public void done() {
```

Comment:
It's to remove the duplication in `RecordBatch` that meant that we had the right behaviour for callbacks, but the wrong behaviour for the return value. If it doesn't affect the tests, it probably means that we are are missing tests for the callbacks case too (which were correct before this PR).

To expand a bit, the call sequence is `future.value()` -> `timestamp()` -> `result.timestamp()`. The latter is set by the `produceFuture.set` call. I'll add a comment to make this clearer.

I hope to have time to add the missing tests at some point this week.